### PR TITLE
Allow the server to cache database objects

### LINKED ISSUE #1050: Cache database objects
This seems like a good idea to reduce latency so that we don’t have to hit PostgreSQL each time, but a new configuration key would be present:

database.caching

Allows to configure the TTL for objects and which strategy (Redis, in-memory) to use.

### PR DESCRIPTION
**Pull request by auguwu**
*closes #1050*

### CODE REVIEW

**Comment by auguwu** [comment]
It seems to work, but no extensive tests exist yet... I'll probably do that later

File: crates/database/src/controller/users.rs [hunk 1]
```diff
@@ -64,8 +84,22 @@ impl DbController for UserDatabaseController {
...
+                    return Ok(Some(cached));
+                }
+
+                self.get(uid).await
+            }
+
+            // we can't do cache calucations on Names since we don't need
+            // duplication, but maybe a "pointer" to their ID?
+            //
+            // TODO(@auguwu): determine if pointers (point name -> id) to a resource is acceptable
```
Comment (spotlightishere) [inline crates/database/src/controller/users.rs:102]: Would it be worth splitting this into an issue so we can track it as the design evolves?

File: crates/search/elasticsearch/src/lib.rs
Comment (spotlightishere) [inline crates/search/elasticsearch/src/lib.rs:1]: I think this is probably unrelated to this pull request, but what caused this to be removed and commented out?
Would it make sense to comment it out like with `melisearch` and the `SearchOptions` definition in `lib.rs`?

File: crates/config/src/caching.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    InMemoryCacheConfig {
+        /// TTL for objects to be discarded in-memory, by default, this will always
+        /// be 15 minutes.
+        #[serde(default, skip_serializing_if = "Option::is_none")]
+        pub time_to_live: Option<Duration> {
+            default: None;
+            env_value: var!("CHARTED_CACHING_REDIS_TTL", to: Duration, is_optional: true);
+        };
+
+        // TODO(@auguwu): add `max_object_size` with custom unit (de)serializer
```
Comment (spotlightishere) [inline crates/config/src/caching.rs:67]: Could this be made into an issue as well, or tracked as part of a goal somewhere?

**Review by spotlightishere — APPROVED** [review]
This appears to be an extremely solid design :) Excited to see this implemented!

I have three general comments related to other changes!

File: crates/database/src/controller/users.rs (hunk 1 above)
Comment (auguwu) [reply crates/database/src/controller/users.rs:102]: Yeah, I think an issue can be created about this

File: crates/search/elasticsearch/src/lib.rs
Comment (auguwu) [reply crates/search/elasticsearch/src/lib.rs:1]: API for it was an old hacky way, wanted to redesign it a tad bit (in a different issue)

File: crates/config/src/caching.rs (hunk 2 above)
Comment (auguwu) [reply crates/config/src/caching.rs:67]: Yes!

**Pull request merged**